metaflow/plugins/kubernetes/kubernetes_decorator.py

```diff
@@ -391,14 +407,80 @@ def _save_package_once(cls, flow_datastore, package):
                 [package.blob], len_hint=1
             )[0]
 
+    @classmethod
+    def _parse_decorator_spec(cls, deco_spec: str):
+        if not deco_spec:
+            return cls()
+
+        valid_options = "|".join(cls.defaults.keys())
+        deco_spec_parts = []
+        for part in re.split(""",(?=[\s\w]+[{}]=)""".format(valid_options), deco_spec):
+            name, val = part.split("=", 1)
+            if name in {"labels", "node_selector"}:
+                try:
+                    tmp_vals = json.loads(val.strip().replace('\\"', '"'))
+                    for val_i in tmp_vals.values():
+                        if not (val_i is None or isinstance(val_i, str)):
+                            raise KubernetesException(
+                                "All values must be string or null."
+                            )
+                except json.JSONDecodeError:
+                    if val.startswith("{"):
+                        raise KubernetesException(
+                            "Malform json detected in %s" % str(val)
+                        )
+                    both = name == "node_selector"
+                    val = json.dumps(
+                        cls.parse_kube_keyvalue_list(val.split(","), both),
+                        separators=(",", ":"),
+                    )
+            deco_spec_parts.append("=".join([name, val]))
+        deco_spec_parsed = ",".join(deco_spec_parts)
+        return super()._parse_decorator_spec(deco_spec_parsed)
+
     @staticmethod
-    def parse_node_selector(node_selector: list):
+    def parse_kube_keyvalue_list(items: List[str], requires_both: bool = True):
         try:
-            return {
-                str(k.split("=", 1)[0]): str(k.split("=", 1)[1])
-                for k in node_selector or []
-            }
+            ret = {}
+            for item_str in items:
+                item = item_str.split("=", 1)
+                if requires_both:
+                    item[1]  # raise IndexError
+                if str(item[0]) in ret:
+                    raise KubernetesException("Duplicate key found: %s" % str(item[0]))
```

Comment:
In its current form the exception does not  convey where the duplicate key was found, as it makes no mention of 'labels' when I set duplicate labels.

As this is used for checks other than the labels as well, A more suitable place to perform the duplicate label check might be in the `validate_kube_labels` instead. That way the exception can be more specific as well (Duplicate *label* found)

Reply:
This function is used on both labels and node selectors and putting the check here ensures that we catch duplicate keys in both sets.